### PR TITLE
Parse case expr

### PR DESCRIPTION
**Pull request by am357**
*Issue #, if available:* #119, #108

*Description of changes:*

Adds CASE expression parsing

GitHub Issue: #119

Adds the required constructs to enable CASE EXPRESSION parsing. The PR includes changes to `partiql-ast` and `partiql-parser` to handle`SearchedCase` and `SimpleCase` parsing.

See the mentioned GitHub issue for more details.

As a result of the changes in this PR, the conformance tests in the following pass:
https://github.com/partiql/partiql-tests/blob/main/partiql-test-data/pass/parser/primitives/case.ion

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: partiql-parser/src/lexer.rs [hunk 1]
```diff
@@ -592,8 +600,8 @@ pub enum Token<'input> {
     Using,
     #[regex("(?i:Value)")]
     Value,
-    #[regex("(?i:Values)")]
-    Values,
```
Comment (jpschorr) [inline partiql-parser/src/lexer.rs:596]: Looks like `Values` was accidentally removed

File: partiql-parser/src/lexer.rs [hunk 2]
```diff
@@ -687,12 +698,13 @@ impl<'input> fmt::Display for Token<'input> {
             | Token::Preserve
             | Token::Right
             | Token::Select
+            | Token::Then
             | Token::True
             | Token::Union
             | Token::Unpivot
             | Token::Using
             | Token::Value
-            | Token::Values
```
Comment (jpschorr) [inline partiql-parser/src/lexer.rs:695]: `Values` removed

File: partiql-parser/src/lexer.rs [hunk 3]
```diff
@@ -720,7 +732,7 @@ mod tests {
             "WiTH Where Value uSiNg Unpivot UNION True Select right Preserve pivoT Outer Order Or \
              On Offset Nulls Null Not Natural Missing Limit Like Left Lateral Last Join \
              Intersect Is Inner In Having Group From Full First False Except Escape Desc \
-             Cross By Between At As And Asc All Values";
```
Comment (jpschorr) [inline partiql-parser/src/lexer.rs:723]: same

File: partiql-parser/src/lexer.rs [hunk 4]
```diff
@@ -740,7 +752,8 @@ mod tests {
             "LIMIT", "/", "LIKE", "^", "LEFT", ".", "LATERAL", "||", "LAST", ":", "JOIN",
             "--", "INTERSECT", "/**/", "IS", "<ident:IDENT>", "INNER", "<atident:@IDENT>", "IN",
             "HAVING", "GROUP", "FROM", "FULL", "FIRST", "FALSE", "EXCEPT", "ESCAPE", "DESC",
-            "CROSS", "BY", "BETWEEN", "AT", "AS", "AND", "ASC", "ALL", "VALUES"
```
Comment (jpschorr) [inline partiql-parser/src/lexer.rs:743]: same

File: partiql-parser/src/parse/partiql.lalrpop [hunk 5]
```diff
@@ -91,16 +91,6 @@ pub SingleQuery: ast::QuerySetAst = {
         }
     },
     <lo:@L> <sfw:SfwQuery> <hi:@R> => ast::QuerySet::Select( Box::new(sfw) ).ast(lo..hi),
-    <lo:@L> <values:Values> <hi:@R> => values,
```
Comment (jpschorr) [inline partiql-parser/src/parse/partiql.lalrpop:94]: Looks like `Values` was accidentally removed

File: partiql-parser/src/parse/partiql.lalrpop [hunk 6]
```diff
@@ -1063,12 +1083,13 @@ extern {
         "PRESERVE" => lexer::Token::Preserve,
         "RIGHT" => lexer::Token::Right,
         "SELECT" => lexer::Token::Select,
+        "THEN" => lexer::Token::Then,
         "TRUE" => lexer::Token::True,
         "UNION" => lexer::Token::Union,
         "UNPIVOT" => lexer::Token::Unpivot,
         "USING" => lexer::Token::Using,
         "VALUE" => lexer::Token::Value,
-        "VALUES" => lexer::Token::Values,
```
Comment (jpschorr) [inline partiql-parser/src/parse/partiql.lalrpop:1071]: `VALUES` removed

File: partiql-parser/src/parse/partiql.lalrpop [hunk 7]
```diff
@@ -739,43 +729,72 @@ ExprPrecedence02: ast::Expr = {
...
+    <lo:@L> "WHEN" <first:ExprQuery> "THEN" <second:ExprQuery> <hi:@R> => ast::ExprPair { first, second },
+};
 
 pub ExprTerm: ast::Expr = {
     "(" <q:Query> ")" => *q,
     <lo:@L> <lit:Literal> <hi:@R> => ast::Expr{ kind: ast::ExprKind::Lit( lit.ast(lo..hi) ) },
     <lo:@L> <path:PathExpr> <hi:@R> => ast::Expr{ kind: ast::ExprKind::Path( path.ast(lo..hi) ) },
-    <ExprTermArray>,
-    <ExprTermBag>,
-    <ExprTermTuple>,
```
Comment (jpschorr) [inline partiql-parser/src/parse/partiql.lalrpop:750]: This looks accidentally reverted.

File: partiql-parser/src/parse/partiql.lalrpop [hunk 8]
```diff
@@ -739,43 +729,72 @@ ExprPrecedence02: ast::Expr = {
...
+// the following:
+// <subquery> ::= <left paren> <query expression> <right paren>
+
+
+CaseExpr: ast::CaseAst = {
+    <lo:@L> "CASE" <expr:ExprQuery?> <cases:ExprPairWhenThen+> <elsexpr:ElseClause?> "END" <hi:@R> => {
+        match expr {
+            None => ast::Case {
+                kind: ast::CaseKind::SearchedCase(
+                    ast::SearchedCase { cases: Vec::new(), default: elsexpr }
```
Comment (jpschorr) [inline partiql-parser/src/parse/partiql.lalrpop:766]: s/`Vec::new()`/`cases`/

File: partiql-parser/src/parse/partiql.lalrpop [hunk 9]
```diff
@@ -739,43 +729,72 @@ ExprPrecedence02: ast::Expr = {
...
+                )
+            }.ast(lo..hi),
+            Some(expr) => ast::Case {
+                kind: ast::CaseKind::SimpleCase(
+                    ast::SimpleCase { expr, cases: Vec::new(), default: elsexpr }
+                )
```
Comment (jpschorr) [inline partiql-parser/src/parse/partiql.lalrpop:772]: s/`Vec::new()`/`cases`/

**Review by jpschorr — COMMENTED** [review]
I think a rebase removed the changes from https://github.com/partiql/partiql-lang-rust/pull/117

Also a small oversight in creation of the `SearchedCase`/`SimpleCase`.

File: partiql-parser/src/lexer.rs (hunk 1 above)
Comment (am357) [reply partiql-parser/src/lexer.rs:596]: Good catch, not sure how it's happened, will restore.

File: partiql-parser/src/parse/partiql.lalrpop (hunk 8 above)
Comment (am357) [reply partiql-parser/src/parse/partiql.lalrpop:766]: Will fix.

File: partiql-parser/src/lexer.rs (hunk 1 above)
Comment (am357) [reply partiql-parser/src/lexer.rs:596]: Looks like the related commit hasn't been picked up.

**Review by jpschorr — APPROVED** [review]

**Pull request merged**